libswc/seat-ws.c

```diff
@@ -0,0 +1,460 @@
+{
+	struct seat *seat = wl_container_of(listener, seat, keyboard_focus_listener);
+	struct event *ev = data;
+	struct input_focus_event_data *event_data = ev->data;
+
+	if (ev->type != INPUT_FOCUS_EVENT_CHANGED)
+		return;
+
+	if (event_data->new) {
+		struct wl_client *client = wl_resource_get_client(event_data->new->surface->resource);
+
+		/* Offer the selection to the new focus. */
+		data_device_offer_selection(seat->base.data_device, client);
+	}
+}
+
+static void
+handle_data_device_event(struct wl_listener *listener, void *data)
+{
+	struct seat *seat = wl_container_of(listener, seat, data_device_listener);
+	struct event *ev = data;
+
+	if (ev->type != DATA_DEVICE_EVENT_SELECTION_CHANGED)
+		return;
+
+	if (seat->base.keyboard->focus.client)
+		data_device_offer_selection(seat->base.data_device, seat->base.keyboard->focus.client);
+}
+
+static void
+handle_swc_event(struct wl_listener *listener, void *data)
+{
+	struct seat *seat = wl_container_of(listener, seat, swc_listener);
+	struct event *ev = data;
+
+	switch (ev->type) {
+	case SWC_EVENT_DEACTIVATED:
+		seat->ignore = true;
+		keyboard_reset(seat->base.keyboard);
+		break;
+	case SWC_EVENT_ACTIVATED:
+		seat->ignore = false;
+		break;
+	}
+}
+
+/* Wayland Seat Interface */
+static void
+get_pointer(struct wl_client *client, struct wl_resource *resource, uint32_t id)
+{
+	struct seat *seat = wl_resource_get_user_data(resource);
+
+	pointer_bind(&seat->pointer, client, wl_resource_get_version(resource), id);
+}
+
+static void
+get_keyboard(struct wl_client *client, struct wl_resource *resource, uint32_t id)
+{
+	struct seat *seat = wl_resource_get_user_data(resource);
+
+	keyboard_bind(seat->base.keyboard, client, wl_resource_get_version(resource), id);
+}
+
+static void
+get_touch(struct wl_client *client, struct wl_resource *resource, uint32_t id)
+{
+	/* XXX: Implement */
+}
+
+static struct wl_seat_interface seat_impl = {
+	.get_pointer = get_pointer,
+	.get_keyboard = get_keyboard,
+	.get_touch = get_touch,
+};
+
+static void
+bind_seat(struct wl_client *client, void *data, uint32_t version, uint32_t id)
+{
+	struct seat *seat = data;
+	struct wl_resource *resource;
+
+	if (version > 4)
+		version = 4;
+
+	resource = wl_resource_create(client, &wl_seat_interface, version, id);
+	wl_resource_set_implementation(resource, &seat_impl, seat, &remove_resource);
+	wl_list_insert(&seat->resources, wl_resource_get_link(resource));
+
+	if (version >= 2)
+		wl_seat_send_name(resource, seat->name);
+
+	wl_seat_send_capabilities(resource, seat->capabilities);
+}
+
+static int
+ws_to_xkb(unsigned type, int key)
+{
+	switch (type) {
+	case WSKBD_TYPE_PC_XT:
+	case WSKBD_TYPE_PC_AT:
+		return wsXtMap[key];
+	case WSKBD_TYPE_USB:
+	case WSKBD_TYPE_MAPLE:
+		return wsUsbMap[key];
+#if 0
+	case WSKBD_TYPE_ADB:
+		return wsAdbMap[key];
+	case WSKBD_TYPE_LK201:
+		return wsLk201Map[key];
+	case WSKBD_TYPE_SUN5:
+	case WSKBD_TYPE_SUN:
+		return wsSunMap[key];
+#endif
```

Comment:
Should this be removed?

Reply:
Yeah, I'll remove it. It's not useful until someone develops a direct rendering video driver for some obscure old workstations... and even then, we should probably map all of this to use USB scancodes in the kernel.